### ISSUE TITLE
Spark session creation is flaky in GitHub Actions

### ISSUE THREAD

**Issue by harupy** [mlflow/mlflow#4229]
https://github.com/mlflow/mlflow/blob/8e74cdd108ff324175e38d621b12d9b2aced05bd/tests/pyfunc/test_spark.py#L56-L67

`get_spark_session` sometimes fails in GitHub Actions with the following error:

```
tests/pyfunc/test_spark.py:66: in get_spark_session
    .master("local-cluster[2, 1, 1024]")
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/pyspark/sql/session.py:173: in getOrCreate
    sc = SparkContext.getOrCreate(sparkConf)
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/pyspark/context.py:349: in getOrCreate
    SparkContext(conf=conf or SparkConf())
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/pyspark/context.py:118: in __init__
    conf, jsc, profiler_cls)
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/pyspark/context.py:180: in _do_init
    self._jsc = jsc or self._initialize_context(self._conf._jconf)
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/pyspark/context.py:288: in _initialize_context
    return self._jvm.JavaSparkContext(jconf)
/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/py4j/java_gateway.py:1525: in __call__
    answer, self._gateway_client, None, self._fqn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

answer = 'xro65'
gateway_client = <py4j.java_gateway.GatewayClient object at 0x7f5d460c2828>
target_id = None, name = 'org.apache.spark.api.java.JavaSparkContext'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.
    
        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.
    
        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
                    raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
E                   : java.net.BindException: Cannot assign requested address: Service 'sparkMaster' failed after 16 retries (on a random free port)! Consider explicitly setting the appropriate binding address for the service 'sparkMaster' (for example spark.driver.bindAddress for SparkDriver) to the correct binding address.
E                   	at sun.nio.ch.Net.bind0(Native Method)
E                   	at sun.nio.ch.Net.bind(Net.java:461)
E                   	at sun.nio.ch.Net.bind(Net.java:453)
E                   	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:222)
E                   	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:128)
E                   	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:558)
E                   	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1283)
E                   	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:501)
E                   	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:486)
E                   	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:989)
E                   	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:254)
E                   	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:364)
E                   	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
E                   	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
E                   	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
E                   	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
E                   	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
E                   	at java.lang.Thread.run(Thread.java:748)

answer     = 'xro65'
gateway_client = <py4j.java_gateway.GatewayClient object at 0x7f5d460c2828>
name       = 'org.apache.spark.api.java.JavaSparkContext'
target_id  = None
type       = 'r'
value      = JavaObject id=o65

/usr/share/miniconda/envs/test-environment/lib/python3.6/site-packages/py4j/protocol.py:328: Py4JJavaError
---------------------------- Captured stdout setup -----------------------------
2021-04-06 05:07:33 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 WARN  Utils:66 - Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-06 05:07:34 ERROR SparkContext:91 - Error initializing SparkContext.
```

# Logs:

[logs_22914.zip](https://github.com/mlflow/mlflow/files/6264429/logs_22914.zip)

**Comment by harupy** [comment]
Related code in spark:

https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/util/Utils.scala#L2310

**Comment by harupy** [comment]
This issue seems related to GitHub Actions.

**Comment by harupy** [comment]
Addressed via #4227
